**Customers Segmentation with RFM Analysis <br> Lucky Wijaya Pengestu**

In [1]:
# Import Library
import pandas as pd 
import datetime as dt 

In [2]:
# Read Dataset
df = pd.read_excel('superstore.xls')
pd.set_option('display.max_columns', None)
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region       Product ID         Category Sub-Category  \
0        42420  South  FUR-BO-10001798        Furniture    Bookcases   
1        42420  South  FUR-CH-10000454        Furniture       Chairs   
2        90036   West  OFF-LA-10000240  Office Supplies       Labels   
3        33311  South  FUR-TA-10000577        Furniture       Tables   
4        33311  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164

In [3]:
# Check DataFrame info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

In [4]:
# Changing Column Name
df.rename(columns = {'Row ID' : 'Row', 'Order ID' : 'OrderID', 'Order Date' : 'OrderDate', 'Ship Date' : 'ShipDate', 
                     'Ship Mode' : 'ShipMode', 'Customer ID' : 'ID', 'Customer Name' : 'Name', 'Postal Code' : 'PostalCode', 
                     'Product ID' : 'ProductID', 'Sub-Category' : 'SubCategory', 'Product Name' : 'Product'},  inplace = True)

In [5]:
df.head(1)

Row         OrderID  OrderDate   ShipDate      ShipMode        ID  \
0    1  CA-2016-152156 2016-11-08 2016-11-11  Second Class  CG-12520   

          Name   Segment        Country       City     State  PostalCode  \
0  Claire Gute  Consumer  United States  Henderson  Kentucky       42420   

  Region        ProductID   Category SubCategory  \
0  South  FUR-BO-10001798  Furniture   Bookcases   

                             Product   Sales  Quantity  Discount   Profit  
0  Bush Somerset Collection Bookcase  261.96         2       0.0  41.9136

# RFM

## Recency

In [6]:
#Last date when data was taken
print(df['OrderDate'].max()) 
now = dt.datetime(2017,12,30) 

2017-12-30 00:00:00


In [7]:
recency = df.groupby(['ID'])['OrderDate'].max().reset_index() #Last order date per customer
recency.columns = ['ID', 'LastOrderDate'] #Rename columns
recency['Recency'] = recency['LastOrderDate'].apply(lambda x: (now - x).days) #Count recency based on LastOrderDate
recency

ID LastOrderDate  Recency
0    AA-10315    2017-06-29      184
1    AA-10375    2017-12-11       19
2    AA-10480    2017-04-15      259
3    AA-10645    2017-11-05       55
4    AB-10015    2016-11-10      415
..        ...           ...      ...
788  XP-21865    2017-11-17       43
789  YC-21895    2017-12-26        4
790  YS-21880    2017-12-21        9
791  ZC-21910    2017-11-06       54
792  ZD-21925    2017-06-11      202

[793 rows x 3 columns]

## Frequency

In [8]:
frequency = df.groupby(['ID'])['OrderID'].count().reset_index() #Count amount of order per customer
frequency.columns = ['ID','Frequency'] #Rename column
frequency

ID  Frequency
0    AA-10315         11
1    AA-10375         15
2    AA-10480         12
3    AA-10645         18
4    AB-10015          6
..        ...        ...
788  XP-21865         28
789  YC-21895          8
790  YS-21880         12
791  ZC-21910         31
792  ZD-21925          9

[793 rows x 2 columns]

## Monetary

In [9]:
monetary = df.groupby(['ID']).agg({'Sales' : 'sum'}).reset_index() #Sum total purchase per customer
monetary.columns = ['ID', 'Monetary'] #Rename column
monetary

ID  Monetary
0    AA-10315  5563.560
1    AA-10375  1056.390
2    AA-10480  1790.512
3    AA-10645  5086.935
4    AB-10015   886.156
..        ...       ...
788  XP-21865  2374.658
789  YC-21895  5454.350
790  YS-21880  6720.444
791  ZC-21910  8025.707
792  ZD-21925  1493.944

[793 rows x 2 columns]

# RFM Analysis

In [10]:
# Merge Recency, Frequency & Monetary columns
rf = recency.merge(frequency, on = 'ID')
rfm = rf.merge(monetary, on = 'ID')
rfm

ID LastOrderDate  Recency  Frequency  Monetary
0    AA-10315    2017-06-29      184         11  5563.560
1    AA-10375    2017-12-11       19         15  1056.390
2    AA-10480    2017-04-15      259         12  1790.512
3    AA-10645    2017-11-05       55         18  5086.935
4    AB-10015    2016-11-10      415          6   886.156
..        ...           ...      ...        ...       ...
788  XP-21865    2017-11-17       43         28  2374.658
789  YC-21895    2017-12-26        4          8  5454.350
790  YS-21880    2017-12-21        9         12  6720.444
791  ZC-21910    2017-11-06       54         31  8025.707
792  ZD-21925    2017-06-11      202          9  1493.944

[793 rows x 5 columns]

In [11]:
# RFM Boundaries Score based on Quartiles
quartiles = rfm.quantile(q = [0.25,0.5,0.75])
quartiles.to_dict()

# Highest recency is the worse value while the lowest recency is the best
def Rscore(value, r, quartiles):
    if value <= quartiles[r][0.25]:
        return 4
    elif value <= quartiles[r][0.50]:
        return 3
    elif value <= quartiles[r][0.75]: 
        return 2
    else:
        return 1
    
# Highest frequency & monetary is the best value while the lowest is the worse
def FMscore(value, fm, quartiles):
    if value <= quartiles[fm][0.25]:
        return 1
    elif value <= quartiles[fm][0.50]:
        return 2
    elif value <= quartiles[fm][0.75]: 
        return 3
    else:
        return 4

In [12]:
# RFM Score Columns
rfm['R_Quartile'] = rfm['Recency'].apply(Rscore, args = ('Recency',quartiles,))
rfm['F_Quartile'] = rfm['Frequency'].apply(FMscore, args = ('Frequency',quartiles,))
rfm['M_Quartile'] = rfm['Monetary'].apply(FMscore, args = ('Monetary',quartiles,))
rfm

ID LastOrderDate  Recency  Frequency  Monetary  R_Quartile  \
0    AA-10315    2017-06-29      184         11  5563.560           1   
1    AA-10375    2017-12-11       19         15  1056.390           4   
2    AA-10480    2017-04-15      259         12  1790.512           1   
3    AA-10645    2017-11-05       55         18  5086.935           3   
4    AB-10015    2016-11-10      415          6   886.156           1   
..        ...           ...      ...        ...       ...         ...   
788  XP-21865    2017-11-17       43         28  2374.658           3   
789  YC-21895    2017-12-26        4          8  5454.350           4   
790  YS-21880    2017-12-21        9         12  6720.444           4   
791  ZC-21910    2017-11-06       54         31  8025.707           3   
792  ZD-21925    2017-06-11      202          9  1493.944           1   

     F_Quartile  M_Quartile  
0             2           4  
1             3           1  
2             2           2  
3             4           4  
4             1           1  
..          ...         ...  
788           4           3  
789           1           4  
790           2           4  
791           4           4  
792           2           2  

[793 rows x 8 columns]

In [13]:
# RFMscore union & total RFM Score
rfm['RFMScore'] = rfm.R_Quartile.map(str) + rfm.F_Quartile.map(str) + rfm.M_Quartile.map(str)
rfm['TotalRFM'] = rfm.R_Quartile + rfm.F_Quartile + rfm.M_Quartile
rfm

ID LastOrderDate  Recency  Frequency  Monetary  R_Quartile  \
0    AA-10315    2017-06-29      184         11  5563.560           1   
1    AA-10375    2017-12-11       19         15  1056.390           4   
2    AA-10480    2017-04-15      259         12  1790.512           1   
3    AA-10645    2017-11-05       55         18  5086.935           3   
4    AB-10015    2016-11-10      415          6   886.156           1   
..        ...           ...      ...        ...       ...         ...   
788  XP-21865    2017-11-17       43         28  2374.658           3   
789  YC-21895    2017-12-26        4          8  5454.350           4   
790  YS-21880    2017-12-21        9         12  6720.444           4   
791  ZC-21910    2017-11-06       54         31  8025.707           3   
792  ZD-21925    2017-06-11      202          9  1493.944           1   

     F_Quartile  M_Quartile RFMScore  TotalRFM  
0             2           4      124         7  
1             3           1      431         8  
2             2           2      122         5  
3             4           4      344        11  
4             1           1      111         3  
..          ...         ...      ...       ...  
788           4           3      343        10  
789           1           4      414         9  
790           2           4      424        10  
791           4           4      344        11  
792           2           2      122         5  

[793 rows x 10 columns]

In [14]:
# Customers Segmentation based on RFMscore
segmentation = []

for index, column in rfm.iterrows():
    if column['TotalRFM'] <= 5:
        segment = 'low'
    elif column['TotalRFM'] > 9 :
        segment = 'high'
    else:
        segment = 'medium'
    segmentation.append(segment)
    
rfm['Segmentation'] = segmentation
rfm

ID LastOrderDate  Recency  Frequency  Monetary  R_Quartile  \
0    AA-10315    2017-06-29      184         11  5563.560           1   
1    AA-10375    2017-12-11       19         15  1056.390           4   
2    AA-10480    2017-04-15      259         12  1790.512           1   
3    AA-10645    2017-11-05       55         18  5086.935           3   
4    AB-10015    2016-11-10      415          6   886.156           1   
..        ...           ...      ...        ...       ...         ...   
788  XP-21865    2017-11-17       43         28  2374.658           3   
789  YC-21895    2017-12-26        4          8  5454.350           4   
790  YS-21880    2017-12-21        9         12  6720.444           4   
791  ZC-21910    2017-11-06       54         31  8025.707           3   
792  ZD-21925    2017-06-11      202          9  1493.944           1   

     F_Quartile  M_Quartile RFMScore  TotalRFM Segmentation  
0             2           4      124         7       medium  
1             3           1      431         8       medium  
2             2           2      122         5          low  
3             4           4      344        11         high  
4             1           1      111         3          low  
..          ...         ...      ...       ...          ...  
788           4           3      343        10         high  
789           1           4      414         9       medium  
790           2           4      424        10         high  
791           4           4      344        11         high  
792           2           2      122         5          low  

[793 rows x 11 columns]

In [15]:
# Counting amount of customers based on segmentation
rfm.groupby(['Segmentation']).agg({'ID' : 'count'}).reset_index()

Segmentation   ID
0         high  200
1          low  202
2       medium  391

**Conclusion**

In determining the performance of customer purchases, we use 3 parameters: <br>
1. **Recency**: Number of days from last order until the last date data was taken <br>
2. **Frequency**: Total orders amount <br>
3. **Monetary**: Total amount that the customer has spent on order <br>

The range of score values for every parameter is 1 to 4. The higher value means better performance. <br>
From the combination of these three parameters, we will create RFM score. From each parameter, the RFM score will be added up to get the total RFM, the total RFM value is used as segmentation for customers whose performance is low, medium, high.

**Segmentation** <br>
TotalRFM <= 5 : Low <br>
TotalRFM 6-9 : Medium <br>
TotalRFM > 9 : High 
<br><br>
With explanation as follows:<br>
Min TotalRFM 3 (combined RFM score: 111)<br>
Max TotalRFM 12 (combined RFM Score: 444)

The low limit is five because the customer has to have a minimum value of 2 points in each parameter or has more value in one of the parameters. <br>
The high limit is nine because there is a possibility that the customer has a total value of 9 with a combination of 144. This means that the customer often makes purchases with large total orders. However, the customer has not repurchased for a long time. <br>
Customers can have a total value of nine with a combination of 414 which means that the customer has just purchased with a large nominal amount. However, the customer has only made a purchase once so he has not become a loyal customer.

The RFM value uses the 25%, 50%, 75% percentile to see the distribution of the data on each parameter and divides it into 4 parts to determine the high/low parameter values owned by each customer. Example: A customer has RFMScore 344 and TotalRFM 11. It means that the customer has a value of Recency 3, Frequency 4, and Monetary 4. This means that the customer has recently placed an order. These customers often place orders and with a large total nominal. If the RFMScore is added, it will total 11 points. This indicates that the customer is included in the high segmentation because it has a total score of more than 9 points.